### PR TITLE
feat(kad): emit `ToSwarm::NewExternalAddrOfPeer`

### PR DESCRIPTION
**Pull request by PanGan21**
## Description

Updates `libp2p-kad` to emit new event `ToSwarm::NewExternalAddrOfPeer` whenever it discovers a new address through the DHT.
Related: #5103 

## Change checklist

- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/kad/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.46.1 - unreleased
```
Comment (guillaumemichel) [inline protocols/kad/CHANGELOG.md]:
```suggestion
## 0.46.1
```

**Review by guillaumemichel — APPROVED** [review]
Looks good to me, thanks @PanGan21 !

File: protocols/kad/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ## 0.46.1
 
+- Emit `ToSwarm::NewExternalAddrOfPeer`.
+  See [PR 5549](https://github.com/libp2p/rust-libp2p/pull/5549)
```
Comment (dariusc93) [inline protocols/kad/CHANGELOG.md:4]: We should move this to 0.46.2 and bump the version in Cargo.toml :)

Comment (PanGan21) [reply protocols/kad/CHANGELOG.md:4]: Done!

**Comment by Wiezzel** [comment]
@PanGan21 Will this emit `NewExternalAddrOfPeer` even when the new peer address is **not** added to the routing table?

**Comment by PanGan21** [comment]
> NewExternalAddrOfPeer

Hi @Wiezzel , I believe it will not. In which cases would this be required?

**Comment by Wiezzel** [comment]
> Hi @Wiezzel , I believe it will not. In which cases would this be required?

Let's say you would like to discover an addr of a peer (e.g. to establish a connection). So, you start a `getClosestPeers` query. As the peer is found, it turns out that the appropriate bucket is full, so it will not be added. In such case, despite having a successful query, and actually discovering the peer's address, we won't receive the event. This seems like a quite common and reasonable case to me.

**Comment by guillaumemichel** [comment]
Let's continue the discussion at https://github.com/libp2p/rust-libp2p/issues/5103

**Pull request merged**